### ISSUE TITLE
Exceptions around EmbedHostObject

### ISSUE THREAD

**Issue by tamaskaroly** [Taritsyn/JavaScriptEngineSwitcher#74]
Hi,

First of all, thanks for your great work! I have an issue using EmbedHostObject, I quite often get unhandled exceptions ("A callback was made on a garbage collected delegate"). I'm using .NET Core 2.2 on Linux (Ubuntu 18.04). The NuGet packet version is 3.1.2 The issue doesn't pop up in the development environment (VS 2019). I know this issue has been discussed before, but I can't find any solution. Here is what I'm trying to do:

_consoleMapper = new ConsoleMapper();
_engine.EmbedHostObject("Console", _consoleMapper);

private class ConsoleMapper
{
  public void write(string text)
  {
    ...
  }
}

Calling Console.write from JS succeeds for a few times, then I get the exception above on the host side. I tried to pin the _consoleMapper object with GCHandle.Alloc, but it didn't help.

Could you please give some guidelines, tips how to properly use EmbedHostObject?

Thanks,
Tamás

**Comment by Taritsyn** [comment]
Hello, Tamás!

Give an example of the full error message with stack trace.

**Comment by tamaskaroly** [comment]
Wow, that was quick! :) Here is the stack trace:

FailFast:
A callback was made on a garbage collected delegate of type 'JavaScriptEngineSwitcher.ChakraCore!JavaScriptEngineSwitcher.ChakraCore.JsRt.JsSerializedLoadScriptCallback::Invoke'.

   at JavaScriptEngineSwitcher.ChakraCore.JsRt.NativeMethods.JsGetAndClearExceptionWithMetadata(JavaScriptEngineSwitcher.ChakraCore.JsRt.JsValue ByRef)
   at JavaScriptEngineSwitcher.ChakraCore.JsRt.NativeMethods.JsGetAndClearExceptionWithMetadata(JavaScriptEngineSwitcher.ChakraCore.JsRt.JsValue ByRef)
   at JavaScriptEngineSwitcher.ChakraCore.JsRt.JsErrorHelpers.ThrowIfError(JavaScriptEngineSwitcher.ChakraCore.JsRt.JsErrorCode)
   at JavaScriptEngineSwitcher.ChakraCore.JsRt.JsValue.CallFunction(JavaScriptEngineSwitcher.ChakraCore.JsRt.JsValue[])
   at JavaScriptEngineSwitcher.ChakraCore.ChakraCoreJsEngine+<>c__DisplayClass28_0.<InnerCallFunction>b__0()
   at JavaScriptEngineSwitcher.ChakraCore.ScriptDispatcher+<>c__DisplayClass10_0`1[[System.__Canon, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].<Invoke>b__0()
   at JavaScriptEngineSwitcher.ChakraCore.ScriptDispatcher.StartThread()
   at System.Threading.Thread.ThreadMain_ThreadStart()
   at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)

**Comment by Taritsyn** [comment]
> I tried to pin the _consoleMapper object with GCHandle.Alloc, but it didn't help.

It's unnecessary.

Give an full source code of the example.

**Comment by Taritsyn** [comment]
And try to upgrade to version 3.1.3.

**Comment by Taritsyn** [comment]
How do you work with the pre-compiled scripts?

First of all, you need to keep the pre-compiled script from garbage collection. Save it in the field or call the `GC.KeepAlive(precompiledScript)` at the end of program. A lot depends on your architecture.

**Comment by tamaskaroly** [comment]
Upgrading to 3.1.3 didn't solve the issue. I also saved the pre-compiled scipt in a field, but it didn't help either. Below is the class that exhibits the issue. Basically, it is used in an ASP.NET app that allows callers to upload a script, then query/update variable values via the rest API.

using JavaScriptEngineSwitcher.ChakraCore;
using JavaScriptEngineSwitcher.Core;
using Microsoft.Extensions.Logging;
using Newtonsoft.Json;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net;
using System.Runtime.InteropServices;
using System.Threading;

namespace DynamicUIManager
{
    public class JavaScriptEngine : IScriptEngine
    {
        private readonly IScriptEngineHostServices _hostServices;
        private readonly ILoggerFactory _loggerFactory;
        private IJsEngine _engine;
        private IPrecompiledScript _precompiledScript;
        private int _initialized;
        private string _sourceCode;
        private ConsoleMapper _consoleMapper;
        private DmccMapper _dmccMapper;

        public JavaScriptEngine(IScriptEngineHostServices hostServices, ILoggerFactory loggerFactory)
        {
            _hostServices = hostServices;
            _loggerFactory = loggerFactory;
        }

        public void Initialize(string sourceCode)
        {
            if (Interlocked.CompareExchange(ref _initialized, 1, 0) != 0)
            {
                return;
            }

            IJsEngineSwitcher engineSwitcher = JsEngineSwitcher.Current;

            engineSwitcher.EngineFactories.AddChakraCore();
            engineSwitcher.DefaultEngineName = engineSwitcher.EngineFactories.First().EngineName;

            _engine = engineSwitcher.CreateDefaultEngine();

            _consoleMapper = new ConsoleMapper(this, _hostServices);
            _dmccMapper = new DmccMapper(this, _hostServices);

            _engine.EmbedHostObject("Console", _consoleMapper);
            _engine.EmbedHostObject("Dmcc", _dmccMapper);

            _sourceCode = sourceCode;
        }

        public void ExecuteScript()
        {
            _precompiledScript = _engine.Precompile(_sourceCode);
            _engine.Execute(_precompiledScript);
        }

        public void Dispose()
        {
            Dispose(true);
            GC.SuppressFinalize(this);
        }

        public object GetVariableValue(string variableName)
        {
            var value = _engine.GetVariableValue(variableName);

            // If the variable is a complex type, get its value as a JSON string and serialize it back to a .NET type.
            if (value != null && value.GetType().FullName == "JavaScriptEngineSwitcher.ChakraCore.JsRt.JsValue")
            {
                var json = _engine.Evaluate<string>($"JSON.stringify({variableName});");

                if (!string.IsNullOrEmpty(json) && json.First() == '[' && json.Last() == ']')
                {
                    value = JsonConvert.DeserializeObject<Newtonsoft.Json.Linq.JArray>(json);
                }
                else
                {
                    value = JsonConvert.DeserializeObject<Newtonsoft.Json.Linq.JObject>(json);
                }
            }

            return value;
        }

        public bool HasVariable(string variableName)
        {
            return _engine.HasVariable(variableName);
        }

        public bool SetVariableValue(string variableName, object variableValue)
        {
            if (!_engine.HasVariable(variableName))
            {
                return false;
            }

            // If the incoming variable is a complex type (JSON), we have to use eval() to inject the value, because
            // the script engine doesn't support complext types.
            if (variableValue != null && (variableValue is Newtonsoft.Json.Linq.JObject || variableValue is Newtonsoft.Json.Linq.JArray))
            {
                _engine.Evaluate($"eval({variableName} = {variableValue.ToString()});");
            }
            else if (variableValue is long)
            {
                _engine.SetVariableValue(variableName, Convert.ToInt32(variableValue));
            }
            else
            {
                _engine.SetVariableValue(variableName, variableValue);
            }

            return true;
        }

        public void CallVariablesUpdatedFunction(string pageId)
        {
            var param = new object[] { pageId };
            var handle = GCHandle.Alloc(param);

            try
            {
                _engine.CallFunction("onVariablesUpdated", param);
            }
            finally
            {
                handle.Free();
            }
        }

        protected virtual void Dispose(bool disposing)
        {
            if (disposing)
            {
                if (_engine != null)
                {
                    _engine.Dispose();
                    _engine = null;
                }
            }
        }

        private class ConsoleMapper
        {
            private readonly IScriptEngine _scriptEngine;
            private readonly IScriptEngineHostServices _hostServices;

            public ConsoleMapper(IScriptEngine scriptEngine, IScriptEngineHostServices hostServices)
            {
                _scriptEngine = scriptEngine;
                _hostServices = hostServices;
            }

            public void write(string text)
            {
                _hostServices.WriteLog(_scriptEngine, text);
            }
        }

        private class DmccMapper
        {
            private readonly IScriptEngine _scriptEngine;
            private readonly IScriptEngineHostServices _hostServices;

            public DmccMapper(IScriptEngine scriptEngine, IScriptEngineHostServices hostServices)
            {
                _scriptEngine = scriptEngine;
                _hostServices = hostServices;
            }

            public int openConnection(string deviceIp, int port, string userName, string password)
            {
                if (!IPAddress.TryParse(deviceIp, out var ip))
                {
                    throw new ArgumentException($"Format of parameter {nameof(deviceIp)} is invalid.");
                }

                var connectionId = _hostServices.OpenDmccConnection(ip, port, userName, password);

                return connectionId;
            }

            public void closeConnection(int connectionId)
            {
                _hostServices.CloseDmccConnection(connectionId);
            }

            public string sendCommand(int connectionId, string command)
            {
                var result = _hostServices.SendDmccCommand(connectionId, command);

                return JsonConvert.SerializeObject(result);
            }
        }
    }
}

**Comment by Taritsyn** [comment]
```csharp
public void ExecuteScript()
{
    _precompiledScript = _engine.Precompile(_sourceCode);
    _engine.Execute(_precompiledScript);
}
```

This code doesn't make sense. Re-read the [“How to upgrade applications to version 3.X > Script pre-compilation”](https://github.com/Taritsyn/JavaScriptEngineSwitcher/wiki/How-to-upgrade-applications-to-version-3.X#script-pre-compilation) subsection of the documentation.

**Comment by Taritsyn** [comment]
```csharp
var handle = GCHandle.Alloc(param);
```

This is unnecessary.

**Comment by Taritsyn** [comment]
```charp
IJsEngineSwitcher engineSwitcher = JsEngineSwitcher.Current;

engineSwitcher.EngineFactories.AddChakraCore();
engineSwitcher.DefaultEngineName = engineSwitcher.EngineFactories.First().EngineName;

_engine = engineSwitcher.CreateDefaultEngine();
```

You can replace all this code with one line:

```csharp
_engine = new ChakraCoreJsEngine();
```

**Comment by Taritsyn** [comment]
```charp
_hostServices.WriteLog(_scriptEngine, text);
```

Why do you call a JS engine from JS engine?

**Comment by Taritsyn** [comment]
It may be better to use the original [Jint](https://github.com/sebastienros/jint) or [Jurassic](https://github.com/paulbartrum/jurassic) libraries for your task?

**Comment by tamaskaroly** [comment]
I just pass on the log entry to the host. That should be ok.

Sometimes I get the following exception. I guess it is the same as the previous one. It clearly shows that the issue is with CallVariablesUpdatedFunction.

JavaScriptEngineSwitcher.Core.JsFatalException: A fatal exception has occurred in a JavaScript runtime

Engine name: ChakraCoreJsEngine
Engine version: 1.11.11
Category: Fatal error
Description: A fatal exception has occurred in a JavaScript runtime ---> JavaScriptEngineSwitcher.ChakraCore.JsRt.JsFatalException: A fatal exception has occurred in a JavaScript runtime
   at JavaScriptEngineSwitcher.ChakraCore.JsRt.JsErrorHelpers.ThrowIfError(JsErrorCode error)
   at JavaScriptEngineSwitcher.ChakraCore.JsRt.JsValue.CallFunction(JsValue[] arguments)
   at JavaScriptEngineSwitcher.ChakraCore.ChakraCoreJsEngine.<>c__DisplayClass28_0.<InnerCallFunction>b__0()
   at JavaScriptEngineSwitcher.ChakraCore.ChakraCoreJsEngine.<>c__DisplayClass28_0.<InnerCallFunction>b__0()
   at JavaScriptEngineSwitcher.ChakraCore.ScriptDispatcher.<>c__DisplayClass10_0`1.<Invoke>b__0()
   at JavaScriptEngineSwitcher.ChakraCore.ScriptDispatcher.StartThread()
--- End of stack trace from previous location where exception was thrown ---
   at JavaScriptEngineSwitcher.ChakraCore.ScriptDispatcher.InnnerInvoke(Func`1 del)
   at JavaScriptEngineSwitcher.ChakraCore.ScriptDispatcher.Invoke[T](Func`1 func)
   at Cognex.DynamicUIManager.JavaScriptEngine.CallVariablesUpdatedFunction(String pageId) in D:\Work\GitRepositories\deadpool\Projects\RTM\Cognex.DynamicUIManager\JavaScriptEngine.cs:line 184
   at Cognex.DynamicUIManager.Server.Controllers.WizardsController.UpdateVariablesForForm(String wizardName, String formName, IEnumerable`1 variables) in D:\Work\GitRepositories\deadpool\Projects\RTM\Cognex.DynamicUIManager.Server\Controllers\WizardsController.cs:line 247

**Comment by Taritsyn** [comment]
I recommend that you follow my previous advice.

**Comment by tamaskaroly** [comment]
I finally figured this out. Actually, it's very simple. The ASP.NET server serves requests on arbitrary threads and the UI made two calls simultaneously. It seems the engine expects requests on the thread it was created on, or at least it is not thread-safe. I created a separate thread that created the engine and then interacts with it and now it's all good.

Thank you for your effort, I really appreciate.
Tamas

**Comment by tamaskaroly** [comment]
Hi,

Thank you for your input.

This is part of a server that initializes an instance of my class once. Some code is standard boilerplate code, some others reflect my struggle to make it work. I’ll clean it up. I removed the _precompiledScript and pass the source code string directly to the engine. ExecuteScript is called once.

Looking at the ScriptDispatcher, I implemented a very similar functionality that all engine calls are dispatched to a thread. I also create the engine instance on that thread, I don’t know if you do the same, but it seems it is the key.

Thanks again for your help, I really appreciate!
Tamás
